viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -186,6 +197,36 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
     return grpc::Status::OK;
 }
 
+int SLAMServiceImpl::GetInternalStateC(
+    viam_carto_get_internal_state_response *response) {
+    boost::uuids::uuid uuid = boost::uuids::random_generator()();
+    std::string filename = path_to_map + "/" + "temp_internal_state_" +
+                           boost::uuids::to_string(uuid) + ".pbstream";
+    {
+        std::lock_guard<std::mutex> lk(map_builder_mutex);
+        bool ok = map_builder.SaveMapToFile(true, filename);
+        if (!ok) {
+            return VIAM_CARTO_SAVE_STATE_ERROR;
+        }
+    }
+
+    std::string buf;
+    // deferring reading the pbstream file in chunks until we run into issues
+    // with loading the file into memory
+    try {
+        ConvertSavedMapToStream(filename, &buf);
+    } catch (std::runtime_error &e) {
+        LOG(ERROR) << e.what();
+        return VIAM_CARTO_SERIALIZATION_ERROR;
```

Comment:
If we don't want to catch all `runtime_errors`, we can change the type of exception thrown just thought we would want to change as little as possible for now, but let me know if you think we should be throwing more specific errors

Reply:
If we needed to handle different runtime errors with different return codes we could look into creating a class per runtime error, however what you have here is appropriate for our current use case.